### ISSUE TITLE
[iOS] canvas.GetStringSize() not considering newlines

### ISSUE THREAD

**Issue by riccardominato** [dotnet/maui#18572]
### Description

In iOS `canvas.GetStringSize()` seems not to handle correctly newlines. It creates a `Size` wide as the longest row and high as a single row.

You can verify this by running the project in repro and setting `NEWLINE_IOS` global variable (if not already set) in MyDrawable.cs.

In my opinion this issue could be linked to https://github.com/dotnet/maui/issues/8486.

### Steps to Reproduce

1. Create a `GraphicsView` with a `Drawable`.
2. Define a string `StringToVisualize` with newlines, like this one: `"HELLO, WORLD!\nCiao mondo row 2\n\nGuten tag!?àèìòù@"`
3. Choose a `FontSize`
4. Inside the `Draw()` method, call `canvas.GetStringSize(StringToVisualize, Font.Default, FontSize);`
5. Draw a rectangle with the size returned by the call at point 4.

Expected Result (Android)
![Screenshot_1699369953](https://github.com/dotnet/maui/assets/69031646/65a6053f-b193-442e-be8e-32a742638b3a)

Actual Result (iOS)
![Simulator Screenshot - iPhone 15 - 2023-11-07 at 16 15 15](https://github.com/dotnet/maui/assets/69031646/be2eb5f9-69e2-47f5-bbde-661bd03e786d)

### Link to public reproduction project repository

https://github.com/Riccardo11/MauiIssues/tree/main/MeasureStringWrongBounds

### Version with bug

8.0.0-rc.2.9511

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

iOS16, iOS17. Couldn't test other versions

### Did you find any workaround?

Calculating string size by hand.

**Comment by drasticactions** [comment]
https://github.com/dotnet/maui/blob/baa86b20786f75632dd99ae13df7d268a5b843f9/src/Graphics/src/Graphics/Platforms/iOS/PlatformStringSizeService.cs#L23-L36

I think this code is wrong. I believe it will _only_ run on iOS and/or catalyst versions of "14.0" and will ignore newer versions, which will cause the fallback to run that doesn't work. Updating it to use "OperatingSystem.IsIOSVersionAtLeast" should fix this, as that actual code in the block should work correctly.

**Comment by riccardominato** [comment]
Don't know if something has been changed in .NET iOS, but in Xamarin.IOS `UIDevice.CheckSystemVersion()` checks

> whether the system version is greater than or equal to the specified major and minor values.

https://learn.microsoft.com/en-us/dotnet/api/uikit.uidevice.checksystemversion?view=xamarin-ios-sdk-12

**Comment by drasticactions** [comment]
@riccardominato Yeah, my explanation was a little off. I'll rephrase:

 "UIDevice.CurrentDevice.CheckSystemVersion(14, 0)" is true, since the version for iOS is above 14. This code block should run. The negation makes it so it will _only_ run below that version, which is also wrong since I believe it wouldn't run correctly below it. So the code being hit is the fallback, which is incorrect. It also won't work correctly on Catalyst, which should use the OperatingSystem check as it does everywhere else in MAUI.

**Comment by riccardominato** [comment]
@drasticactions Thank you for your explanation. So if I understand correctly the problem is that negation, which shouldn't be there.

However, I can't see why it checks on iOS 14, because [`nsString.StringSize` is deprecated since iOS 7.0](https://developer.apple.com/documentation/foundation/nsstring/1619910-sizewithfont). Afaik MAUI only supports iOS 11 and newer, so maybe no checks are required.

**Comment by drasticactions** [comment]
I believe the first use of that code would predate MAUI, either in MAUI Graphics or in some form of SkiaSharp support for Forms, which could explain the older UIKit API check that was attempted. I agree though looking at the supported OS versions for MAUI, this check isn't needed. I've introduced a PR to remove it.

**Comment by riccardominato** [comment]
Thank you for taking action so fast. I had a look to the pull request and I tried to simulate the behavior MAUI will have after merging it. This is the code:

```csharp
#if IOS
    var cgSize = new NSString(StringToVisualize).GetBoundingRect( 
        CGSize.Empty, 
        NSStringDrawingOptions.UsesLineFragmentOrigin,
        new UIStringAttributes { Font = Font.Default.ToPlatformFont(FontSize),  }, 
        null).Size;
    var stringSize = new SizeF((float)cgSize.Width, (float)cgSize.Height);
#endif
```

and this is the result:

<img width="176" alt="Screenshot 2023-11-08 alle 12 03 10" src="https://github.com/dotnet/maui/assets/69031646/1d4e7f8c-bc55-4bb7-9af9-1be1ee69ca4d">

The screenshot you posted in your pull request seems to calculate better the bounding box: did you test it on my repro or you made some edits?
Does it have something to do with [your comment](https://github.com/dotnet/maui/pull/18621#issuecomment-1801604083) about Skia drawing?

**Comment by riccardominato** [comment]
Ok, it seems there's something wrong in general with calculating strings bounding box.

It draws a smaller rectangle even with a single line string, regardless the method used (deprecated or not).

<img width="211" alt="Screenshot 2023-11-08 alle 12 47 04" src="https://github.com/dotnet/maui/assets/69031646/b7bb1931-a85c-403a-8f99-b1a30ee17833">

**Comment by drasticactions** [comment]
@riccardominato Yes, I think that this could be issues in general with how this API is designed.

If you set a specific Font (Ex. `new Font("Ariel")` then it should get a closer result to what you expect (That's where my screenshot came from, without it I get your result) but I think that's because this calculation from UIKit does not reflect what the actual drawing size will be once rendered with Skia. If you literally use that NSString or NSAttributedString in a UILabel, that size should be correct. But with the drawing coming from elsewhere (In this case, Skia), that is not reflected.

So this is more approximate of the size, rather than it being totally true, IMO. But considering that the current code is calling deprecated APIs anyway, IMO, it's a good first step to at least call the right thing as it is written. Once that lands, then we can figure out a better way to handle that method.

@mattleibow as you'll know more about this, is what I'm saying sound true? Do you have thoughts on this?

**Comment by riccardominato** [comment]
@drasticactions I agree on your PR commits being the right first step to solve this issue, but I'm honestly a bit concerned about your proposal.

Waiting for the PR to be merged, published in a release and certifying that there is a problem we already know it's there seems a long time period and carries the risk that the real problem will be forgotten.

I don't know how much time consuming this activity would be for you, but IMO it would be more reasonable to tackle the real problem instead of publishing a "better version to get wrong" and postponing the real solution to the future.

Thank you for your support.

**Comment by drasticactions** [comment]
@riccardominato I don't work on the MAUI Team. I'm here helping when I can, generally focusing on elements of this platform that need help to support hot reloading, which is a team I do work for.

My PR will run the correct API as it should have from the start, and will get this codepath running on the correct API and not a deprecated one. It may still have other issues, but it's better than it was before. If anyone can make a better fix or has other ideas, they can PR it. If MAUI team members rather have a different fix, they can reject it. For me, I'm trying to manage my time to help where I can.

**Comment by riccardominato** [comment]
@drasticactions thank you for your support and for your PR. I'm going to open another issue to describe the problem that emerged after going in deep in this issue.
